### ISSUE TITLE
audiod_tx_done_cb overflow

### ISSUE THREAD

**Issue by HiFiPhile** [hathach/tinyusb#935]
### Operating System

Windows 10

### Board

Core4357

### Firmware

examples\device\uac2_headset

### What happened ?

Playing with uac2_headset I found there are output samples in the buffer, but mic input is always silence.

Initially there are IN xfer, but as soon as OUT intf is opened, IN xfer stops.
```
00>   AUDIO xfer callback
00>   Queue EP 81 with 0 bytes ... OK
00> USBD Xfer Complete on EP 81 with 0 bytes
00>   AUDIO xfer callback
00>   Queue EP 81 with 0 bytes ... OK
00> USBD Xfer Complete on EP 81 with 0 bytes
00>   AUDIO xfer callback
00>   Queue EP 81 with 0 bytes ... OK
00> USBD Xfer Complete on EP 81 with 0 bytes
00>   AUDIO xfer callback
00>   Queue EP 81 with 0 bytes ... OK
00> USBD Xfer Complete on EP 81 with 0 bytes
00>   AUDIO xfer callback
00>   Queue EP 81 with 0 bytes ... OK
00> USBD Xfer Complete on EP 81 with 0 bytes
00>   AUDIO xfer callback
00> 
00> USBD Setup Received 01 0B 01 00 01 00 00 00 
00>   Set Interface
00>   AUDIO control request
00>   Set itf: 1 - alt: 1
00>   Open EP 01 with Size = 192
00> Set interface 1 alt 1
00>   Queue EP 01 with 192 bytes ... OK
00>   Queue EP 80 with 0 bytes ... OK
00> USBD Xfer Complete on EP 80 with 0 bytes
00> 
00> USBD Setup Received 01 0B 00 00 01 00 00 00 
00>   Set Interface
00>   AUDIO control request
00>   Set itf: 1 - alt: 0
00>   CLOSING Endpoint: 0x01
00>   Queue EP 80 with 0 bytes ... OK
00> USBD Xfer Complete on EP 80 with 0 bytes
00> 
00> USBD Setup Received 01 0B 01 00 01 00 00 00 
00>   Set Interface
00>   AUDIO control request
00>   Set itf: 1 - alt: 1
00>   Open EP 01 with Size = 192
00> Set interface 1 alt 1
00>   Queue EP 01 with 192 bytes ... OK
00>   Queue EP 80 with 0 bytes ... OK
00> USBD Xfer Complete on EP 80 with 0 bytes
00> USBD Xfer Complete on EP 01 with 192 bytes
00>   AUDIO xfer callback
00>   Queue EP 01 with 192 bytes ... OK
00> USBD Xfer Complete on EP 01 with 192 bytes
00>   AUDIO xfer callback
00>   Queue EP 01 with 192 bytes ... OK
00> USBD Xfer Complete on EP 01 with 192 bytes
00>   AUDIO xfer callback
```
So I looked inside, added a log inside `audiod_tx_done_cb`:
```TU_LOG2("Tx done Cb idxItf:%d  alt:%d\r\n", idxItf, audio->alt_setting[idxItf]);```
I got:
```
00>   AUDIO xfer callback
00> Tx done Cb idxItf:6  alt:17
00>   Queue EP 81 with 0 bytes ... OK
00> USBD Xfer Complete on EP 81 with 0 bytes
00>   AUDIO xfer callback
00> Tx done Cb idxItf:6  alt:81
00>   Queue EP 81 with 0 bytes ... OK
00> USBD Xfer Complete on EP 81 with 0 bytes
00>   AUDIO xfer callback
00> Tx done Cb idxItf:6  alt:209

```
@PanRe  Could you take a look at `audio->alt_setting[idxItf]`, since in this example `alt_setting_1` has a length of 1, access 6th element is not normal.

### How to reproduce ?

1. Run `examples\device\uac2_headset`
2. Check `audio->alt_setting[idxItf]` in `audiod_tx_done_cb`.

### Debug Log
[3.txt](https://github.com/hathach/tinyusb/files/6733054/3.txt)

**Comment by PanRe** [comment]
I am a bit puzzled how this come to happen, but before digging deep into the code a few questions:
- Did you use the fix #849?
- Did you try disabling the debug messages? In my setup, streaming does not work at all if i enable the debug messages

If nothing helps, we need to debug `audiod_get_AS_interface_index()`...

**Comment by PanRe** [comment]
> You have different return in your board ?

I never tested this example on my own. I only adapted the changes required for the new version of the UAC2 driver.

Looking at the example in more detail i see that there are two AS interfaces defined in the audio descriptors, yet `CFG_TUD_AUDIO_FUNC_1_N_AS_INT = 1` is defined within `tusb_config.h`. It should be `CFG_TUD_AUDIO_FUNC_1_N_AS_INT = 2` yielding `alt_setting_1` to be of length two. If interface 1 is opened, `idxItf = 0` should be the result of `audiod_get_AS_interface_index().` Likewise, for interface 2 it should be `idxItf = 1`.

Could you verify the descriptor boarders `audio->p_desc` and `p_desc_end` within `audiod_get_AS_interface_index()`?

**Comment by HiFiPhile** [comment]
I think it's a logic problem on `tmp` increment, `tmp` should increase only on alt0 interface descriptor, now it's good :)
```
    uint8_t tmp = 0;
    while (p_desc < p_desc_end)
    {
      // We assume the number of alternate settings is increasing thus we return the index of alternate setting zero!
      if (tu_desc_type(p_desc) == TUSB_DESC_INTERFACE && ((tusb_desc_interface_t const * )p_desc)->bAlternateSetting == 0)
      {
        if (((tusb_desc_interface_t const * )p_desc)->bInterfaceNumber == itf)
        {
          *idxItf = tmp;
          *pp_desc_int = p_desc;
          return true;
        }
        // Increase index, bytes read, and pointer
        tmp++;
      }
      p_desc = tu_desc_next(p_desc);
    }
```

**Comment by PanRe** [comment]
Bingo, that should it be! Thanks for fixing this bug... my working example while developing that stuff did not cover this situation! I hope there are not many other possibilities i did not cover :D

Is the example working now?

**Comment by HiFiPhile** [comment]
> Is the example working now?

Yes it works now.  

I'll also add `feature_unit_set_request` for `UAC2_ENTITY_CLOCK`, change frequency is quite common.

**Comment by PanRe** [comment]
Jep feel free to enhance it! Change of sample resolution from 16/24/32 bit may also be nice but is not mandatory. Some time ago i tried to do this, but i did not manage it to get 24 bits working on Windows. The standard UAC2 Windows driver ignored my 24 bit alternate interface setting for some reason i don't know yet. Have you done this before?

**Comment by HiFiPhile** [comment]
I haven't tried 24bit yet.
For EP `wMaxPacketSize` it can be figured out more smartly with maximum sample frequency, for example with xmos:
```
/* Max packet sizes:
 * Samples per channel. e.g (192000+7999)/8000 = 24
 * Must allow 1 sample extra per chan (24 + 1) = 25
 * Multiply by number of channels and bytes      25 * 2 * 4 = 200 bytes
 * TODO Output doesn't get modified by channel count
*/
#define MAX_PACKET_SIZE_MULT_OUT_HS ((((MAX_FREQ+7999)/8000)+1) * NUM_USB_CHAN_OUT)
#define MAX_PACKET_SIZE_MULT_OUT_FS ((((MAX_FREQ_FS+999)/1000)+1) * NUM_USB_CHAN_OUT_FS)

#define HS_STREAM_FORMAT_OUTPUT_1_MAXPACKETSIZE (MAX_PACKET_SIZE_MULT_OUT_HS * HS_STREAM_FORMAT_OUTPUT_1_SUBSLOT_BYTES)
#define HS_STREAM_FORMAT_OUTPUT_2_MAXPACKETSIZE (MAX_PACKET_SIZE_MULT_OUT_HS * HS_STREAM_FORMAT_OUTPUT_2_SUBSLOT_BYTES)
#define HS_STREAM_FORMAT_OUTPUT_3_MAXPACKETSIZE (MAX_PACKET_SIZE_MULT_OUT_HS * HS_STREAM_FORMAT_OUTPUT_3_SUBSLOT_BYTES)

#if (HS_STEAM_FORMAT_OUPUT_1_MAXPACKETSIZE > 1024) || (HS_STEAM_FORMAT_OUPUT_2_MAXPACKETSIZE > 1024) \
    || (HS_STEAM_FORMAT_OUPUT_3_MAXPACKETSIZE > 1024)
#error
#endif

#define FS_STREAM_FORMAT_OUTPUT_1_MAXPACKETSIZE (MAX_PACKET_SIZE_MULT_OUT_FS * FS_STREAM_FORMAT_OUTPUT_1_SUBSLOT_BYTES)
#define FS_STREAM_FORMAT_OUTPUT_2_MAXPACKETSIZE (MAX_PACKET_SIZE_MULT_OUT_FS * FS_STREAM_FORMAT_OUTPUT_2_SUBSLOT_BYTES)
#define FS_STREAM_FORMAT_OUTPUT_3_MAXPACKETSIZE (MAX_PACKET_SIZE_MULT_OUT_FS * FS_STREAM_FORMAT_OUTPUT_3_SUBSLOT_BYTES)
```

**Comment by PanRe** [comment]
Mh i tried something like that... i read that the standard windows driver selects the alternate interface whose wMaxPacketSize is just big enough to handle the desired bandwidth. So i had
_(48000 / 1000 + 1) * N_Ch * N_BYTES_PER_SAMPLE_ for _N_BYTES_PER_SAMPLE = 2, 3, 4_
But this did not work so far...
Anyway, thanks for fixing the example!!

**Comment by HiFiPhile** [comment]
Maybe it's because `bInterval`, in this example `bInterval` was 4 which should be 1, only the feedback EP should set `bInterval` to 4.

With `bInterval` set to 1 I can set `wMaxPacketSize` like in theory.

I pushed my changes in #938

**Comment by PanRe** [comment]
Thanks a lot! A working example is always nice to have! Which program did you use to get the descriptors and the nice parsing of them?

**Comment by HiFiPhile** [comment]
> Which program did you use to get the descriptors and the nice parsing of them?

This one : https://www.uwe-sieber.de/usbtreeview_e.html

**Comment by PanRe** [comment]
Thanks a lot! :)

**Comment by HiFiPhile** [comment]
Closed by #938
